internal/util/reftracker/reftracker.go

```diff
@@ -0,0 +1,162 @@
+/*
+Copyright 2022 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reftracker
+
+import (
+	goerrors "errors"
+	"fmt"
+
+	"github.com/ceph/go-ceph/rados"
+
+	"github.com/ceph/ceph-csi/internal/util/reftracker/errors"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/radoswrapper"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/reftype"
+	v1 "github.com/ceph/ceph-csi/internal/util/reftracker/v1"
+	"github.com/ceph/ceph-csi/internal/util/reftracker/version"
+)
+
+// reftracker is key-based implementation of a reference counter.
+//
+// Unlike integer-based counter, reftracker counts references by tracking
+// unique keys. This allows accounting in situations where idempotency must be
+// preserved. It guarantees there will be no duplicit increments or decrements
+// of the counter.
+//
+// It is stored persistently as a RADOS object, and is safe to use with
+// multiple concurrent writers, and across different nodes of a cluster.
+
+// Add atomically adds refs to reference tracker named `rtName`.
+// If the reftracker object doesn't exist yet, it is created. This is also
+// reflected in the return value.
+func Add(
+	ioctx radoswrapper.IOContextW,
+	rtName string,
+	refs map[string]struct{},
+) (created bool, err error) {
+	if err = validateAddInput(rtName, refs); err != nil {
+		return false, err
+	}
+
+	// Read reftracker version.
+
+	rtVer, err := version.Read(ioctx, rtName)
+	if err != nil {
+		if goerrors.Is(err, rados.ErrNotFound) {
+			// This is a new reftracker. Initialize it with `refs`.
+			if err = v1.Init(ioctx, rtName, refs); err != nil {
+				return false, fmt.Errorf("failed to initialize reftracker: %w", err)
+			}
+
+			return true, nil
+		}
+
+		return false, fmt.Errorf("failed to read reftracker version: %w", err)
+	}
+
+	// Add references to reftracker object.
+
+	gen, err := ioctx.GetLastVersion()
+	if err != nil {
+		return false, fmt.Errorf("failed to get RADOS object version: %w", err)
+	}
+
+	switch rtVer {
+	case v1.Version:
+		err = v1.Add(ioctx, rtName, gen, refs)
+		if err != nil {
+			err = fmt.Errorf("failed to add refs: %w", err)
+		}
+	default:
+		err = errors.UnknownObjectVersion(rtVer)
+	}
+
+	return false, err
```

Comment:
should we return `true` in case `v1.Add` is successful?

Reply:
Similary to what I've mentioned in https://github.com/ceph/ceph-csi/pull/2893/files/4f7800a48ff563bc7c6d759bdd7a7f9b2fa5bb2a#r831982758, I hoped that using named returns (`(created bool, err error)`) would help users notice what the returned boolean represents. It returns `true` only if the object was created by the `Add()` call. In other cases (when it already exists, and refs are simply added to it) it's not created, so it must return `false`.

Reply:
I'll try to make the comment clearer though.

Reply:
Yes, that helps. return true only if the object is created not for  adding reference.